man/rdd.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "RDD" "1" "February 2013" "" ""
.
.SH "NAME"
\fBrdd\fR \- Readability command\-line client
.
.SH "SYNOPSIS"
\fBrdd\fR resources [\fIoptions\fR]
.
.P
\fBrdd\fR shorten \fIfull\-url\fR [\fIoptions\fR]
.
.P
\fBrdd\fR metadata \fIurl\-id\fR [\fIoptions\fR]
.
.SH "DESCRIPTION"
\fBrdd\fR is a simple command\-line tool to utilize the Readability Shortener API \fIhttps://www\.readability\.com/developers/api/rdd\fR\.
.
.SH "COMMANDS"
.
.TP
resources
Retrieve information about sub\-resources\.
.
.TP
shorten \fIfull\-url\fR
Create a new shortened URL\.
.
.TP
metadata \fIurl\-id\fR
Retrieve available metadata of a shortened link\.
.
.SH "OPTIONS"
.
.TP
\fB\-h\fR, \fB\-\-help\fR
Show a help message describing all available options\.
.
.TP
\fB\-u\fR, \fB\-\-url\fR
Set URL of Readability shortener service\.
.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Be a bit more verbose\.
.
.TP
\fIfull\-url\fR
Long URL to be shortened, e\.g\. "http://www\.paulgraham\.com/gh\.html"\.
.
.TP
\fIurl\-id\fR
URL ID as returned by the Readability API, e\.g\. "ga4qf47t"\.
.
.SH "ENVIRONMENT"
The following environment variables are evaluated:
.
.TP
\fBRDD_URL\fR
Set URL of Readability shortener service\. Can also be controlled by the \fB\-\-url\fR command\-line option\.
.
.TP
\fBRDD_VERBOSE\fR
Define to enable verbose mode\. Can also be controlled by the \fB\-\-verbose\fR command\-line option\.
.
.SH "EXAMPLES"
.
.nf

$ rdd resources
urls:
  description:
    The URLs endpoint\. POST a URL to add it to the shortener\.
  href:
    /api/shortener/v1/urls
urls/:id:
  description:
    The URL endpoint\. GET a URL ID to view available metadata of a shortened link\.
  href:
    /api/shortener/v1/urls/:id

$ rdd shorten http://www\.paulgraham\.com/gh\.html
id:
  ga4qf47t
rdd_url:
  http://rdd\.me/ga4qf47t
url:
  /api/shortener/v1/urls/ga4qf47t

$ rdd metadata ga4qf47t
article:
  author:
    None
  excerpt:
    Want to start a startup? Get funded by Y Combinator\. \.\.\.
  title:
    Great Hackers
  url:
    http://www\.paulgraham\.com/gh\.html
  word_count:
    5147
full_url:
  http://readability\.com/articles/ga4qf47t
id:
  ga4qf47t
rdd_url:
  http://rdd\.me/ga4qf47t
.
.fi
.
.SH "AUTHOR"
Written by Mathias Lafeldt \fImathias\.lafeldt@gmail\.com\fR


HISTORY.rst
History
-------

0.2.3 (2013-02-18)
++++++++++++++++++

* Use HTTPretty as mock library for unit tests.
* Add commands ``test_units`` and ``test_integration`` to setup.py.
* Update classifiers in setup.py.

0.2.2 (2013-02-15)
++++++++++++++++++

* Fix exit status of ``python setup.py test``.
* Fix path to rdd module in tests.
* Update README.
* Update manpage.
* Update copyright year in LICENSE file.

0.2.1 (2013-02-11)
++++++++++++++++++

* Fix JSON decoding in case no text is returned.
* Sort JSON data to address hash randomization in Python 3.3.
* Travis: run test against Python 3.3.
* Add Travis status to README.

0.2.0 (2012-12-19)
++++++++++++++++++

* Update to Requests version 1.0.3.

0.1.9 (2012-11-06)
++++++++++++++++++

* Add unit tests powered by `pytest <http://pytest.org>`_.
* Work around dependency error returned by ``pip install`` by hardcoding the
  current rdd.py version in setup.py.

0.1.8 (2012-07-23)
++++++++++++++++++

* Only return API data of interest and output that data using own formatter
  instead of json.
* Add dedicated Readability exceptions.
* Add help text to command-line options.
* Update to Sharness v0.2.4.

0.1.7 (2012-06-07)
++++++++++++++++++

* Make use of ``Response.json`` from Requests v0.12.1.
* Add Python 3 support.
* Add Travis CI config, testing Python 2.6, 2.7, and 3.2.
* Update to Sharness v0.2.1.

0.1.6 (2012-02-09)
++++++++++++++++++

* Fix URL of shortener service. For whatever reason, HTTPS is no longer
  supported. Only http://www.readability.com/api/shortener/v1 works.
* Disable annoying HTTP POST redirection.

0.1.5 (2012-01-18)
++++++++++++++++++

* Add more automated tests.
* Fix tests to pass on Mac OS X Lion.
* Set ``RDD_VERBOSE`` with test option -v.

0.1.4 (2011-12-05)
++++++++++++++++++

* Support environment variables ``RDD_URL`` and ``RDD_VERBOSE``.
* Additionally print HTTP content in verbose mode.
* Add rdd.exceptions (imports exceptions from Requests).
* Improve documentation (install instructions, docstrings, etc.)

0.1.3 (2011-12-04)
++++++++++++++++++

* Add automated tests powered by Sharness.

0.1.2 (2011-11-23)
++++++++++++++++++

* Split up rdd.py into package and runner script.
* Add setup.py for easy installation.

0.1.1 (2011-11-22)
++++++++++++++++++

* Add ``--url`` option.
* Allow HTTP redirects.
* Import ``json`` if ``simplejson`` is not available.
* Add ``__author__``, ``__license__``, and ``__version__``.
* Add manpage in ronn format.
* Syntax highlighting in README.
* Add this history file.

0.1.0 (2011-11-21)
++++++++++++++++++

* First version (written on the same day the Readability API was released)


rdd.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Python implementation of the Readability Shortener API

rdd.py can either be imported as a Python module or invoked as a command-line
client.
"""

__author__ = 'Mathias Lafeldt <mathias.lafeldt@gmail.com>'
__license__ = 'MIT'
__version__ = '0.1.1'

import requests
try:
    import simplejson as json
except ImportError:
    import json


class Readability(object):

    def __init__(self, url=None, verbose=None):
        self.url = url or 'https://readability.com/api/shortener/v1'
        self.config = {}
        if verbose is not None:
            self.config['verbose'] = verbose

    def _request(self, method, path, data=None, headers=None):
        url = self.url + path
        r = requests.request(method, url, data=data, headers=headers,
                             config=self.config, allow_redirects=True)
        r.raise_for_status()

        if not 'application/json' in r.headers['Content-Type']:
            raise TypeError('No JSON in response')

        return json.loads(r.content) if r.content.strip() else None

    def resources(self):
        """Retrieve information about sub-resources."""
        return self._request('GET', '/')

    def shorten(self, full_url):
        """Create a new shortened URL."""
        headers = {'Content-Type': 'application/x-www-form-urlencoded'}
        data = 'url=%s' % full_url
        return self._request('POST', '/urls', data=data, headers=headers)

    def metadata(self, url_id):
        """Retrieve available metadata of a shortened link."""
        return self._request('GET', '/urls/%s' % url_id)


if __name__ == '__main__':
    import sys
    import optparse

    def die(msg):
        sys.exit('error: %s' % msg)

    parser = optparse.OptionParser()
    parser.add_option('-u', '--url', action='store')
    parser.add_option('-v', '--verbose', action='store_true')
    opts, args = parser.parse_args()

    if len(args) < 1:
        die('command missing')

    verbose = sys.stderr if opts.verbose else None
    readability = Readability(url=opts.url, verbose=verbose)

    cmd = args[0]
    if cmd in ('resources', 'r'):
        data = readability.resources()
    elif cmd in ('shorten', 's'):
        if len(args) < 2:
            die('argument missing')
        data = readability.shorten(args[1])
    elif cmd in ('metadata', 'm'):
        if len(args) < 2:
            die('argument missing')
        data = readability.metadata(args[1])
    else:
        die('invalid command')

    # HACK re-encode json for pretty output
    print json.dumps(data, indent=4)
